platforms/software/build-init/src/main/java/org/gradle/api/tasks/wrapper/internal/GradleVersionResolver.java

```diff
@@ -0,0 +1,147 @@
+        this.releaseNightly = textFactory.fromUri(versionUrl + "/release-nightly");
+    }
+
+    public GradleVersion getGradleVersion() {
+        if (gradleVersion == null) {
+            gradleVersion = GradleVersion.version(resolve(gradleVersionString));
+        }
+        return gradleVersion;
+    }
+
+    public void setGradleVersionString(String gradleVersionString) {
+        if (!isPlaceHolder(gradleVersionString)) {
+            try {
+                this.gradleVersion = GradleVersion.version(gradleVersionString);
+            } catch (Exception e) {
+                throw new WrapperVersionException("Invalid version specified for argument '--gradle-version'", e);
+            }
+        }
+
+        if (!this.gradleVersionString.equals(gradleVersionString)) {
+            this.gradleVersionString = gradleVersionString;
+            this.gradleVersion = null;
+        }
+    }
+
+    private String resolve(@Nullable String version) {
+        if (version == null) {
+            return GradleVersion.current().getVersion();
+        }
+        switch (version) {
+            case LATEST:
+                return getVersion(latest.asString(), version);
+            case NIGHTLY:
+                return getVersion(nightly.asString(), version);
+            case RELEASE_NIGHTLY:
+                return getVersion(releaseNightly.asString(), version);
+            case RELEASE_CANDIDATE:
+                return getVersion(releaseCandidate.asString(), version);
+            case RELEASE_MILESTONE:
+                return getVersion(releaseMilestone.asString(), version);
+            default:
+                return version;
+        }
+    }
+
+    private static String getVersion(String json, String placeHolder) {
+        Type type = new TypeToken<Map<String, String>>() {}.getType();
+        Map<String, String> map = new Gson().fromJson(json, type);
+        String version = map.get("version");
+        if (version == null) {
+            throw new GradleException("There is currently no version information available for '" + placeHolder + "'.");
+        }
+        return version;
+    }
+
+    private static boolean isPlaceHolder(String version) {
+        return PLACE_HOLDERS.contains(version);
+    }
+
+    /**
+     * This exception is thrown when the wrapper task is run in an attempt to update the wrapper version and
+     * an invalid version is specified.
+     */
+    public static final class WrapperVersionException extends GradleException implements ResolutionProvider {
+        public WrapperVersionException(String message, @Nullable Throwable cause) {
+            super(message, cause);
+        }
+
+        @Override
+        public List<String> getResolutions() {
+            return Arrays.asList(suggestActualVersion(), suggestDynamicVersions());
+        }
+
+        private static String suggestActualVersion() {
+            return "Specify a valid Gradle release listed on https://gradle.org/releases/.";
```

Comment:
💭 Unnecessary, but suggesting the actual latest 1) patch for the current minor, 2) minor of the current major, 3) next major, 4) latest overall; would be pretty cool to do here.